add notes and questions for n45 (#748)
bitcoin-core-review-club/website: _posts/2024-03-13-bitcoin-inquisition-#45.md

  unchanged.

The semantics of the signature-checking opcodes are similar to Tapscript
opcodes, including the empty signature->false behavior for
`OP_CHECKSIGFROMSTACK`.

### [The code](https://github.com/bitcoin-core-review-club/bitcoin/tree/prbitcoin-inquisition-45)

#### `OP_INTERNALKEY`

Because this PR is for bitcoin inquisition, the
[implementation](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1309)
of `OP_INTERNALKEY` is remarkably simple. The taproot internal key is already
extracted for use with verifying tapscript `OP_CHECKSIG(VERIFY|ADD)` with the
key `0x01`, so it simply needs to be added to the stack.

#### `OP_CHECKSIGFROMSTACK(VERIFY)`

The implementation of CSFS(V) is split between the [main opcode
processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).

> My original approach was to add a `SignatureChecker` that could carry the

Reviewer: I still see first person here? You can say "a previous approach."

message rather than generating the message from the transaction, and give CSFS
the exact semantics of existing sigops in their respective script types, but
that approach had two problems: getting BIP340 signing into legacy/segwitv0
scripts became intrusive, and the semantics of legacy/segwitv0 signature
checking are not particularly desirable for a new opcode. As a result I
changed to this implementation that uses a separate function specific to CSFS
and gives modern Tapscript-style semantics. Only after changing code
approaches did I realize that ECDSA checking might not be desirable in these
new sigops. The commit history still reflects the removal of ECDSA in a
separate commit.

Outside of these changes in `interpreter.cpp`, the only non-flag change in
production code is to allow variable length input to BIP340
signature-checking.

### Testing

> For both INTERNALKEY and CSFS I wrote tests in the Transaction Test framework:
> 
> https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671
> https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491

Reviewer: raw links are not very reader-friendly
```suggestion
Test data for INTERNALKEY and CSFS are included in [tx_valid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671) and [tx_invalid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491).
```

You can run a regtest node from this branch to use them yourself.

Here are some scripts using `bitcoinjs-lib` and `noble-curves` that were used
to generate the test vectors above:
https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
1. What does `OP_CHECKSIGFROMSTACKVERIFY` leave on the stack?
   `OP_CHECKSIGFROMSTACK`?
1. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?
1. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?
1. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend
   in weight?

> 

Reviewer: nit: prefer having newlines in between questions

commit:
---

## Notes

Attendees should familiarize themselves with
[LNHANCE](https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376),
and [LN-Symmetry (Eltoo)](https://bitcoinops.org/en/topics/eltoo/).

TL;DR: LNHANCE is a soft fork proposal for bitcoin which combines 4 total
opcodes: `OP_CHECKTEMPLATEVERIFY`, `OP_CHECKSIGFROMSTACK`,

> `OP_CHECKSIGFROMSTACKVERIFY`, and `OP_INTERNALKEY`. It consumes 2 NOPs in all

Reviewer: clarification for myself: by "consumes," I assume you mean repurposes?

processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).

An earlier approach was to add a `SignatureChecker` that could carry the
message rather than generating the message from the transaction, and give CSFS
the exact semantics of existing sigops in their respective script types, but
that approach had two problems: getting BIP340 signing into legacy/segwitv0
scripts became intrusive, and the semantics of legacy/segwitv0 signature
checking are not particularly desirable for a new opcode. As a result the
current implementation uses a separate function specific to CSFS and gives
modern Tapscript-style semantics. Only after changing code approaches did it
become clear that ECDSA checking might not be desirable in these new sigops.
The commit history still reflects the removal of ECDSA in a separate commit.

Outside of these changes in `interpreter.cpp`, the only non-flag change in

> production code is to allow variable length input to BIP340

Reviewer: Confused as to what you mean by "production code." Also I don't think the term "flag change" (and "non-flag change") is defined?

signature-checking.

### Testing

Test data for both CHECKSIGFROMSTACK and INTERNALKEY has been added to
[tx_valid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671)
and
[tx_invalid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491)
for the trasnsaction test framework.
[Here](https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd)
are some scripts using `bitcoinjs-lib` and `noble-curves` that were used to
generate the vectors.

You can run a regtest node from this branch to use them yourself.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

2. What does `OP_CHECKSIGFROMSTACKVERIFY` leave on the stack?
   `OP_CHECKSIGFROMSTACK`?

3. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?

4. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?

5. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend

>    in weight?

Reviewer: Some more question ideas
- Did you run the tests? Is there any missing test coverage? Can you think of ways to extend the tests?
- How might one construct a vault (or one of the examples you mention above)?
- What concrete improvements to LN can be achieved this way?

## Open considerations

> 1. Given callers of `OP_CHECKSIGFROMSTACK` can reasonably be expected to want

Reviewer: This question was confusing for me, as I am missing explanations for:
- why callers would be "reasonably expected to want" this ability
- why `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK` is "the naive implementation"
- why it is insecure (where is this explained?)

Reviewer: My recommendation is to split this into multiple questions and have notes walking us through each assertion you're making

commit:
---

## Notes

Attendees should familiarize themselves with
[LNHANCE](https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376),
and [LN-Symmetry (Eltoo)](https://bitcoinops.org/en/topics/eltoo/).

As Inquisition already contains an implementation of BIP-119
(`OP_CHECKTEMPLATEVERIFY`), implementing LNHANCE on Inquisition only requires
adding implementations of BIN-2024-0003 and -0004.

This Pull Request adds a total of 3 opcodes:
* `OP_INTERNALKEY` in Tapscript only, copies the Taproot Internal Key to the
  stack.
* `OP_CHECKSIGFROMSTACK` in Tapscript only, validates a BIP340 Schnorr
  signature using an item from the stack as the message.
* `OP_CHECKSIGFROMSTACKVERIFY` in all script types, validates a BIP340 Schnorr
  signature using an item from the stack as the message, and leaves the stack
  unchanged.

> The semantics of the signature checking opcodes are similar to Tapscript

Reviewer 2: ```suggestion
The semantics of the signature-checking opcodes are similar to Tapscript
```
silly nit, I know

opcodes, including the empty signature->false behavior for
`OP_CHECKSIGFROMSTACK`.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

## Open considerations

1. Given callers of `OP_CHECKSIGFROMSTACK` can reasonably be expected to want
   the ability to check a signature against multiple stack items, and that the

>    naieve implementation (using `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK`) is

Reviewer 2: ```suggestion
   naive implementation (using `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK`) is
```

commit:
---

## Notes

Attendees should familiarize themselves with
[LNHANCE](https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376),

> and [LN-Symmetry (Eltoo)](https://bitcoinops.org/en/topics/eltoo/).

Reviewer: Maybe describe LNHANCE with a few sentences?

the exact semantics of existing sigops in their respective script types, but
that approach had two problems: getting BIP340 signing into legacy/segwitv0
scripts became intrusive, and the semantics of legacy/segwitv0 signature
checking are not particularly desirable for a new opcode. As a result I
changed to this implementation that uses a separate function specific to CSFS
and gives modern Tapscript-style semantics. Only after changing code
approaches did I realize that ECDSA checking might not be desirable in these
new sigops. The commit history still reflects the removal of ECDSA in a
separate commit.

Outside of these changes in `interpreter.cpp`, the only non-flag change in
production code is to allow variable length input to BIP340
signature-checking.

### Testing

For both INTERNALKEY and CSFS I wrote tests in the Transaction Test framework:

https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671
https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491

You can run a regtest node from this branch to use them yourself.

> I posted some scripts using `bitcoinjs-lib` and `noble-curves` that I used to

Reviewer: Please use third person

generate the test vectors above:
https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

> ## Open considerations

Reviewer: I'd recommend adding some questions to help people check their understanding. Softballs like "which nop is being repurposed" / "what does this opcode do" or conceptual questions "what's the motivation for this change" / "how is this useful" are good places to start. Otherwise the meeting will be very short...
